utxoscanner.go

```diff
@@ -0,0 +1,387 @@
+type Interface interface {
+	GetBlockFromNetwork(chainhash.Hash, ...QueryOption) (*btcutil.Block, error)
+	GetBlockHash(int64) (*chainhash.Hash, error)
+	BestSnapshot() (*waddrmgr.BlockStamp, error)
+	GetCFilter(blockHash chainhash.Hash, filterType wire.FilterType,
+		options ...QueryOption) (*gcs.Filter, error)
+	blockFilterMatches(ro *rescanOptions, blockHash *chainhash.Hash) (bool,
+		error)
+}
+
+// A PriorityQueue implements heap.Interface and holds GetUtxoRequests. The
+// queue maintains that heap.Pop() will always return the GetUtxo request with
+// the least starting height. This allows us to add new GetUtxo requests to
+// an already running batch.
+type PriorityQueue []*GetUtxoRequest
+
+func (pq PriorityQueue) Len() int { return len(pq) }
+
+func (pq PriorityQueue) Less(i, j int) bool {
+	// We want Pop to give us the least StartHeight.
+	return pq[i].StartHeight < pq[j].StartHeight
+}
+
+func (pq PriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+}
+
+// Push is called by the heap.Interface implementation to add an element to the
+// end of the backing store. The heap library will then maintain the heap
+// invariant.
+func (pq *PriorityQueue) Push(x interface{}) {
+	item := x.(*GetUtxoRequest)
+	*pq = append(*pq, item)
+}
+
+// Peek returns the least height element in the queue without removing it.
+func (pq *PriorityQueue) Peek() *GetUtxoRequest {
+	return (*pq)[0]
+}
+
+// Pop is called by the heap.Interface implementation to remove an element from
+// the end of the backing store. The heap library will then maintain the heap
+// invariant.
+func (pq *PriorityQueue) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	*pq = old[0 : n-1]
+	return item
+}
+
+// IsEmpty returns true if the queue has no elements.
+func (pq *PriorityQueue) IsEmpty() bool {
+	return pq.Len() == 0
+}
+
+// UtxoScanner batches calls to GetUtxo so that a single scan can search for
+// multiple outpoints. If a scan is in progress when a new element is added, we
+// check whether it can safely be added to the current batch, if not it will be
+// included in the next batch.
+type UtxoScanner struct {
+	pq      PriorityQueue
+	stopped bool
+
+	chainClient Interface
+
+	nextBatch []*GetUtxoRequest
+
+	cv *sync.Cond
+}
+
+// NewUtxoScanner creates a new instance of UtxoScanner using the given chain
+// interface.
+func NewUtxoScanner(chainClient Interface) UtxoScanner {
+	return UtxoScanner{
+		pq:          make(PriorityQueue, 0),
+		cv:          sync.NewCond(&sync.Mutex{}),
+		stopped:     false,
+		chainClient: chainClient,
+	}
+}
+
+// Start begins running scan batches.
+func (s *UtxoScanner) Start() {
+	go func() {
+		for {
+			s.cv.L.Lock()
+			if s.stopped {
+				s.cv.L.Unlock()
+				return
+			}
+
+			// Re-queue previously skipped requests for next batch.
+			for _, request := range s.nextBatch {
+				heap.Push(&s.pq, request)
+			}
+
+			s.nextBatch = nil
```

Comment:
we should consider using
```golang
s.nextBatch = s.nextBatch[:0]
```
to reuse the backing array

Reply:
on second thought, ignore what i said :P 
it might be useful to release this slice so that this can be garbage collected and downsized automatically. otherwise we may retain more space than necessary after the initial rescan.